src/bot/music.rs

```diff
@@ -59,112 +60,87 @@ pub enum MusicBotMessage {
+    type Result = Result<(), AudioPlayerError>;
 }
 
 pub struct MusicBot {
     name: String,
-    player: Arc<AudioPlayer>,
+    identity: Identity,
+    player: AudioPlayer,
     teamspeak: Option<TeamSpeakConnection>,
-    playlist: Arc<RwLock<Playlist>>,
-    state: Arc<RwLock<State>>,
+    master: Option<WeakAddress<MasterBot>>,
+    playlist: Playlist,
+    state: State,
+    logger: Logger,
 }
 
 pub struct MusicBotArgs {
     pub name: String,
-    pub name_index: usize,
-    pub id_index: usize,
+    pub master: Option<WeakAddress<MasterBot>>,
     pub local: bool,
     pub address: String,
-    pub id: Identity,
+    pub identity: Identity,
     pub channel: String,
     pub verbose: u8,
-    pub disconnect_cb: Box<dyn FnMut(String, usize, usize) + Send + Sync>,
+    pub logger: Logger,
 }
 
 impl MusicBot {
-    pub async fn new(args: MusicBotArgs) -> (Arc<Self>, impl Future<Output = ()>) {
-        let (tx, mut rx) = tokio::sync::mpsc::unbounded_channel();
-        let tx = Arc::new(RwLock::new(tx));
-        let (player, connection) = if args.local {
-            info!("Starting in CLI mode");
-            let audio_player = AudioPlayer::new(tx.clone(), None).unwrap();
-
-            (audio_player, None)
-        } else {
-            info!("Starting in TeamSpeak mode");
-
-            let con_config = Connection::build(args.address)
-                .version(tsclientlib::Version::Linux_3_3_2)
-                .name(format!("🎵 {}", args.name))
-                .identity(args.id)
-                .log_commands(args.verbose >= 1)
-                .log_packets(args.verbose >= 2)
-                .log_udp_packets(args.verbose >= 3)
-                .channel(args.channel);
-
-            let connection = TeamSpeakConnection::new(tx.clone(), con_config)
-                .await
-                .unwrap();
-            let mut cconnection = connection.clone();
-            let audio_player = AudioPlayer::new(
-                tx.clone(),
-                Some(Box::new(move |samples| {
-                    let mut rt = tokio::runtime::Runtime::new().unwrap();
-                    rt.block_on(cconnection.send_audio_packet(samples));
-                })),
-            )
-            .unwrap();
-
-            (audio_player, Some(connection))
-        };
-
+    pub async fn new(args: MusicBotArgs) -> Address<Self> {
+        let mut player = AudioPlayer::new(args.logger.clone()).unwrap();
         player.change_volume(VolumeChange::Absolute(0.5)).unwrap();
-        let player = Arc::new(player);
-        let playlist = Arc::new(RwLock::new(Playlist::new()));
 
-        spawn_gstreamer_thread(player.clone(), tx.clone());
+        let playlist = Playlist::new(args.logger.clone());
 
-        if args.local {
-            spawn_stdin_reader(tx);
-        }
+        let teamspeak = if args.local {
+            info!(args.logger, "Starting in CLI mode");
+            debug!(args.logger, "Spawning stdin reader thread");
```

Comment:
no we're not :-o

Reply:
If we log that later we have to clone the logger (KV map & sender). :/

Reply:
Moved it anyway.